Add the hierarchy to legends

trevan opened the pull request:
This fixes #3402.

It also cleaned up some unused code in legends and made the mouseover/mouseout selectors more specific.

The reason that pieNames() returns the labels in two formats is because the color() function requires an array and I also needed the hierarchy structure.  So instead of creating two very similar functions, I just made pieNames() return both the data required.  I did attempt to convert the hierarchy structure to an array but the order would be different which caused colors to look a little wierd.

rashidkpc commented:
Few issues so far:

1) This causes bar charts with only 1 series to be black:

![screen shot 2015-04-10 at 4 20 17 pm](https://cloud.githubusercontent.com/assets/1250387/7098919/b3b76fee-df9d-11e4-89b0-50e50837f97a.png)

2) Short labels + scrollbar causes clipping on the right

![screen shot 2015-04-10 at 4 19 08 pm](https://cloud.githubusercontent.com/assets/1250387/7098922/c83d03b6-df9d-11e4-8296-8a655c00fafb.png)

3) The amount of offset here seems unnecessary. It seems like these could be offset by a smaller fixed amount instead of by the entire width of the legend value.

![screen shot 2015-04-10 at 4 21 29 pm](https://cloud.githubusercontent.com/assets/1250387/7098925/cc7157f2-df9d-11e4-8261-f1d2ca6db6ba.png)

trevan commented:
I fixed the black color issue.  I hadn't realized that the legend and the actual data would not have the same labels.  I added a flag to the color function to handle this case.

I reduced the padding of the legends.  I haven't been able to see item 2 specifically.  I always see the dots as shown in item 3.

trevan commented:
Also, I added the two extra commits marked as squashable.  Do you want fixup commits squashed or leave them separate?

trevan commented:
How do I restart the tests?  It looks like Kibana didn't start up for some reason.

trevan commented:
I've rebased off of master to handle merge conflicts.

rashidkpc commented on src/kibana/components/vislib/components/color/color.js:
```diff
@@ -9,9 +9,10 @@ define(function (require) {
      */
 
-    return function (arrayOfStringsOrNumbers) {
+    return function (arrayOfStringsOrNumbers, matchEmptyToFirst) {
```
Why is an empty value getting passed in? It seems like thats the problem? Why would we want to return the first thing in the array if we get an empty string?

trevan replied:
When you have a bar chart that only has one series, the individual data points don't have a label.  So it originally worked by generating a fake label just for the legend (https://github.com/elastic/kibana/blob/master/src/kibana/components/vislib/lib/data.js#L45) and creating two color functions: one for the legend (https://github.com/elastic/kibana/blob/master/src/kibana/components/vislib/lib/data.js#L53) and one for the actual data (https://github.com/elastic/kibana/blob/master/src/kibana/components/vislib/lib/data.js#L53).

So the legend has a color function that uses an array with a single string value and the data points have a color function that uses an array with a single empty string value.

I thought about adding a label to all the data points to match the legend but wasn't sure that you would want that extra work

Another thought was to modify where it grabbed the label (https://github.com/elastic/kibana/blob/master/src/kibana/components/vislib/lib/data.js#L53) and if it is blank, set it to the 'yAxisLabel'.

Would you prefer either of those to my current code?  Do you have another suggestion?

rashidkpc commented:
This breaks the current behavior of expansion of truncated legend values on hover,

rashidkpc commented:
The values seem to be sorted the same for each outer ring in the visualization, and sorted the same in the table, but differently in each bucket of the legend:

![screen shot 2015-04-20 at 2 24 06 pm](https://cloud.githubusercontent.com/assets/1250387/7240856/0d639e14-e769-11e4-94ab-fe0bf246c831.png)

trevan commented:
I have the legend sorted by size.  That's how all my pie charts showed as.  So I'll have to see how to generate a pie chart that isn't sorted and look into how to make it not change the sort.

trevan commented:
I fixed the legend truncation issue and the legend sort should now match the table and viz.

trevan commented:
It looks like pull request #3641 is going to cause a lot of conflicts with my changes.  If that pull request is going to go in, it would probably be better for me to restart on top of those changes.  It should make it a bit easier since he is passing the entire data set into the legend.  It looks like I would only need to modify the two new transform data functions and the list function in the legend class.

@rashidkpc, what would you like me to do?

rashidkpc commented:
https://github.com/elastic/kibana/pull/3641 is definitely going in

rashidkpc commented:
He @trevan I really liked where this was going. I'm going to close the issue for now, but please do reopen when you're ready.
